exercises/forth/src/example/kotlin/ForthEvaluator.kt

```diff
@@ -0,0 +1,139 @@
+
+                require(newWordEnd >= 0) { "Incomplete word definition" }
+
+                val newWordToken = Token.fromString(trimmedLine.substring(0 until newWordEnd)).first()
+
+                require(newWordToken is WordToken) { "Cannot redefine numbers" }
+
+                val newWordDefTokens = Token.fromString(trimmedLine.substring(newWordEnd + 1))
+                return WordDefToken((newWordToken as WordToken).word, newWordDefTokens)
+            }
+        }
+
+    }
+
+    data class WordToken(val word: String): Token()
+    data class IntToken(val rawValue: Int): Token()
+
+    companion object {
+        val wordChars = "A-z+/*\\-"
+
+        fun fromString(string: String): List<Token> {
+            return when {
+                string.startsWith(':') -> listOf(WordDefToken.fromString(string))
+                string.matches(Regex("[$wordChars]+(?:-[$wordChars]+)*")) -> listOf(WordToken(string.toLowerCase()))
+                string.matches(Regex("\\d+")) -> listOf(IntToken(string.toInt()))
+                else -> string.split(delimiters = " ").flatMap { Token.fromString(it) }
+            }
+        }
+    }
+
+}
+
+class ForthEvaluator {
+
+    companion object {
+        private val builtInWordsAndOperators = mapOf<String, (Deque<Int>) -> Unit>(
+                Pair("+") { values ->
+                    require(values.size >= 2) { "Addition requires that the stack contain at least 2 values" }
+                    values.push(values.pop() + values.pop())
+                },
+
+                Pair("-") { values ->
+                    require(values.size >= 2) { "Subtraction requires that the stack contain at least 2 values" }
+                    val topValue = values.pop()
+                    val secondValue = values.pop()
+                    values.push(secondValue - topValue)
+                },
+
+                Pair("*") { values ->
+                    require(values.size >= 2) { "Multiplication requires that the stack contain at least 2 values" }
+                    values.push(values.pop() * values.pop())
+                },
+
+                Pair("/") { values ->
+                    require(values.size >= 2) { "Division requires that the stack contain at least 2 values" }
+                    val topValue = values.pop()
+                    val secondValue = values.pop()
+                    require(topValue != 0) { "Division by 0 is not allowed" }
+                    values.push(secondValue / topValue)
+                },
+
+                Pair("dup") { values ->
+                    require(values.isNotEmpty()) { "Duplicating requires that the stack contain at least 1 value" }
+                    values.push(values.peek())
+                },
+
+                Pair("drop") { values ->
+                    require(values.isNotEmpty()) { "Dropping requires that the stack contain at least 1 value" }
+                    values.pop()
+                    Unit
```

Comment:
This feels _super_ weird. I needed something here because `values.pop` returns non-Unit. Is there a better way to do this?